storybooks/storybook
# Addons, core: Make `react` and Storybook packages `devDependencies` where possible

Works on #24490 

## ~TODO~
- [ ] ~Document new `optimizeDeps` preset property~
- [x] Reword "no React" error
- [x] Test that an addon author can import from `manager-api` and still have correct types, even though `@storybook/types` and `@types/react` are not part of the package.






## What I did

This PR is a retry of #23486 with many non-essential things ignored.

The main purpose of this PR is to minimize the amount of peer dependencies in addons and a few of the core packages, most notably removing `react` as a peer dependency.

The way the manager and the preview globalizes a fixed set of dependencies means that these packages don't strictly need the dependencies, as they will be referenced from the global window instead, which our builders ensures. Removing them from peer dependencies mean that users don't need to specify them either.

Having them as devDependencies will however make our (tsup) bundler default to bundle them in, which we don't want. Therefore this PR also introduces a new bundler configuration, `addon-bundle` that makes sure that the globalized packages are externalized (thus the imports are kept intact), so they can be replaced with the global reference when built.
A handful of packages like `manager-api` are _not_ using `addon-bundle` but still need these packages externalized, so they've been added to their bundler configuration manually to ensure that happens.

`addon-essentials` and `addon-docs` are largely left out of this PR, as tackling them in terms of `react` requires more work.



## Checklist for Contributors

### Testing



#### The changes in this PR are covered in the following automated tests:

#### Manual testing

It's important to test this in the different package manager configuration, npm, yarn, yarn pnp and pnpm and with both builders.

For each configuration and builder:

1. Create a new non-React sandbox outside of the monorepo (`npx sb@next sandbox`)
2. Replace all versions with this PR's canary version.

#### Testing `addon-essentials` _with_ docs

3. Run `storybook` and `build-storybook` and see that it works

#### Testing `addon-docs` directly

1. Replace `addon-essentials` with `addon-docs` both in `main.ts` and `package.json`
2. Run `storybook` and `build-storybook` and see that it works

#### Testing `addon-essentials` with docs _disabled_

1. Remove the `mdx` glob from `stories` in `main.ts`
1. Replace `addon-docs` with `addon-essentails` both in `main.ts` and `package.json` again
2. Remove `react` and `react-dom` from `package.json`. You will see an unmet peer deps from `addon-essentials`, that's okay.
3. disable `docs` in `main.ts` with `{ name: '@storybook/addon-essential', options: { docs: false } }`
4. Run `storybook` and `build-storybook` and see that it works, now without React

#### Testing _without_ `addon-essentials` and `addon-docs`

1. Remove `addon-docs` with `addon-essentails` both from `main.ts` and `package.json` again
2. Run `storybook` and `build-storybook` and see that it works



### Documentation



      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release


This pull request has been released as version [`0.0.0-pr-24676-sha-214a6f84`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24676-sha-214a6f84). Install it by pinning all your Storybook dependencies to that version.
<details>
<summary>More information</summary>

| | |
| --- | --- |
| **Published version** | [`0.0.0-pr-24676-sha-214a6f84`](https://npmjs.com/package/@storybook/cli/v/0.0.0-pr-24676-sha-214a6f84) |
| **Triggered by** | @JReinhold |
| **Repository** | [storybookjs/storybook](https://github.com/storybookjs/storybook) |
| **Branch** | [`experiment-addon-peer-dep`](https://github.com/storybookjs/storybook/tree/experiment-addon-peer-dep) |
| **Commit** | [`214a6f84`](https://github.com/storybookjs/storybook/commit/214a6f845fcb3af4e07839ff9288c71f5d6a0e02) |
| **Datetime** | Wed Nov  8 11:20:01 UTC 2023 (`1699442401`) |
| **Workflow run** | [6797494862](https://github.com/storybookjs/storybook/actions/runs/6797494862) |

To request a new release of this pull request, mention the `@storybookjs/core` team.

_core team members can create a new canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=24676`_
</details>
